src/rendergraph/common/rendergraph/geometry.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include <QVector2D>
+#include <QVector3D>
+
+#include "backend/geometry.h"
+#include "rendergraph/attributeset.h"
+
+namespace rendergraph {
+class Geometry;
+} // namespace rendergraph
+
+class rendergraph::Geometry : public rendergraph::backend::Geometry {
+  public:
+    struct Point2D {
+        QVector2D position2D;
+        Point2D(float x, float y)
+                : position2D{x, y} {
+        }
+    };
+
+    struct TexturedPoint2D {
+        QVector2D position2D;
+        QVector2D texcoord2D;
+        TexturedPoint2D(float x, float y, float tx, float ty)
+                : position2D{x, y},
+                  texcoord2D{tx, ty} {
+        }
+    };
+
+    struct RGBColoredPoint2D {
+        QVector2D position2D;
+        QVector3D color3D;
+        RGBColoredPoint2D(float x, float y, float r, float g, float b)
+                : position2D{x, y},
+                  color3D{r, g, b} {
+        }
+    };
+
+    struct RGBAColoredPoint2D {
+        QVector2D position2D;
+        QVector4D color4D;
+        RGBAColoredPoint2D(float x, float y, float r, float g, float b, float a)
+                : position2D{x, y},
+                  color4D{r, g, b, a} {
+        }
+    };
+
+    enum class DrawingMode {
+        Triangles,
+        TriangleStrip
+    };
+
+    Geometry(const rendergraph::AttributeSet& attributeSet, int vertexCount);
+    ~Geometry();
+
+    void allocate(int vertexCount);
+
+    void setAttributeValues(int attributePosition, const float* data, int numTuples);
+
+    float* vertexData();
+
+    template<typename T>
+    T* vertexDataAs() {
+        return reinterpret_cast<T*>(vertexData());
+    }
```

Comment:
not really a fan of this in theory, but I see that its modeled after the QSGGeometry class...

Reply:
no, that's not try. QSGGeometry has explicit methods:
const void *	[vertexData](https://doc.qt.io/qt-6/qsggeometry.html#vertexData-1)() const
QSGGeometry::ColoredPoint2D *	[vertexDataAsColoredPoint2D](https://doc.qt.io/qt-6/qsggeometry.html#vertexDataAsColoredPoint2D)()
const QSGGeometry::ColoredPoint2D *	[vertexDataAsColoredPoint2D](https://doc.qt.io/qt-6/qsggeometry.html#vertexDataAsColoredPoint2D-1)() const
QSGGeometry::Point2D *	[vertexDataAsPoint2D](https://doc.qt.io/qt-6/qsggeometry.html#vertexDataAsPoint2D)()
const QSGGeometry::Point2D *	[vertexDataAsPoint2D](https://doc.qt.io/qt-6/qsggeometry.html#vertexDataAsPoint2D-1)() const
QSGGeometry::TexturedPoint2D *	[vertexDataAsTexturedPoint2D](https://doc.qt.io/qt-6/qsggeometry.html#vertexDataAsTexturedPoint2D)()
const QSGGeometry::TexturedPoint2D *	[vertexDataAsTexturedPoint2D](https://doc.qt.io/qt-6/qsggeometry.html#vertexDataAsTexturedPoint2D-1)() const

But that comes with a shortcoming: If you want to use vertex data that doesn't match any of the predefined types, you have to cast it anyway.

Reply:
yes, My problem is not with the generic `vertexDataAs` (in fact as you rightfully point out, its likely superior to the explicit version). I rather meant that this is not typesafe. You can pass the wrong `T` to the `vertexDataAs` use `setAttributeValues` wrong. Ideally the type information was either attached to the buffer at runtime or encoded in the class. We can still use type-erasure internally (in fact we need to as Qt does), but lets try to create a safe API around it.